quarkchain/cluster/shard_state.py

```diff
@@ -85,6 +86,8 @@ def __init__(self, env, full_shard_id: int, db=None, diff_calc=None):
 
         # new blocks that passed POW validation and should be made available to whole network
         self.new_block_pool = dict()
+        # header hash -> (height, [coinbase address]) during previous blocks (ascending)
+        self.coinbase_addr_cache = dict()  # type: Dict[bytes, Tuple[int, Deque[bytes]]]
```

Comment:
I would suggest using lrucache in https://docs.python.org/3/library/functools.html

Reply:
wrap `functools.lrucache` to the method (`__get_coinbase_addresses_until_block`) is tricky in this use case since we have a rolling window AND we want to limit how recursive the function goes:

`rolling_window(n) = rolling_window(n-1).popleft().append(n)`

in our happy path, the simple dict-based cache should work well since blocks will be appended one by one.

fyi the wrapper method `_get_posw_coinbase_blockcnt` indeed uses `lrucache`